[bug] fixes #1158

git-afsantos opened the pull request:
XmlLoader and LoaderContext no longer share the param list to child 'node' contexts.
This was causing the creation of unintended private parameters when the tilde notation was used.

git-afsantos committed:
[bug] fixes #1158

XmlLoader and LoaderContext no longer share the param list to child 'node' contexts.
This was causing the creation of unintended private parameters when the tilde notation was used.

mikepurvis commented:
Thanks for the contribution! Can you add relevant test cases to this pair of files please:

https://github.com/ros/ros_comm/blob/lunar-devel/tools/roslaunch/test/unit/test_roslaunch_dump_params.py

https://github.com/ros/ros_comm/blob/lunar-devel/tools/roslaunch/test/xml/test-dump-rosparam.launch

git-afsantos committed:
added test cases for tilde param in launch

git-afsantos committed:
added test cases for tilde param in python

git-afsantos committed:
fixed tilde param issue for group tags

Issue #1158 manifested for group tags that appear before (but not containing) node tags.

git-afsantos committed:
added one more test case for issue #1158
used param tag to make sure we test the proposed fix

git-afsantos committed:
Merge pull request #1 from git-afsantos/origin/patch-1

added one more test case for issue #1158

git-afsantos commented:
I believe I have added the necessary test cases now. I just noticed that the issue also manifested in `<group>` tags, so I added a few more commits to fix it. Example:

```xml
<launch>
  <param name="~tilde1" value="foo" />
  <group>
    <param name="~tilde2" value="bar" />
  </group>
  <node pkg="package" type="base_test" name="test_node" />
</launch>
```

Results in

```javascript
{"/test_node/tilde1": "foo", "/test_node/tilde2": "bar"}
```

Clearly `tilde2` should not be added to the node, as it is in a different context.

mikepurvis commented on tools/roslaunch/test/unit/test_roslaunch_dump_params.py:
```diff
@@ -77,6 +77,10 @@ def test_roslaunch(self):
+            '/node_rosparam/tilde1': 'foo',
+            '/node_rosparam/local_param': 'baz',
+
+            '/node_rosparam2/tilde1': 'foo',
```
Is there a way to do a negative test here, since that was the actual problem in the first place? eg:

```
    '/node_rosparam2/tilde2': None
```

Might need something changed in the python file to correctly handle it.

git-afsantos replied:
I think the equality tests (`val == output_val` and `assertEquals(val, output_val)`) already covered that case. However, for clarity, I added the negative tests at the bottom.

git-afsantos committed:
Added negative tests for extra parameters

Some parameters should not be present at all.

mikepurvis commented:
This looks okay to me. @dirk-thomas?

dirk-thomas commented:
Thank you!
